### PR TITLE
Add `LRFinder` callback

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

will add docs and examples in a follow up.

Added `LearningRateFinder` callback that allows users to run lr_finder at any time they want during `.fit` call using a callback.
```py
class CustomLearningRateFinder(LearningRateFinder):
    def on_fit_start(self, trainer, pl_module):
        self.lr_find(trainer, pl_module)
    
    def on_train_epoch_start(self, trainer, pl_module):
        if self.trainer.current_epoch == 10:
            self.lr_find(trainer, pl_module)


lr_cb = CustomLearningRateFinder(...)
trainer = Trainer(..., callbacks=lr_cb)
trainer.fit(...)

```

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @rohitgr7 @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/callbacks/batch_size_finder.py [hunk 1]
```diff
@@ -26,9 +26,41 @@
...
 from pytorch_lightning.utilities.parsing import lightning_hasattr
 from pytorch_lightning.utilities.rank_zero import rank_zero_warn
+from pytorch_lightning.utilities.seed import isolate_rng
 
 
 class BatchSizeFinder(Callback):
+    """The ``BatchSizeFinder`` callback tries to find the largest batch size for a given model that does not give
+    an out of memory (OOM) error. It works with both training and evalation. All you need to do is add it as a
+    callback inside Trainer and call ``trainer.fit/validate/test/predict()``. Internally it calls the respective
+    step function ``steps_per_trial`` times for each batch size until one of the batch size generates and OOM
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:36]:
```suggestion
    step function ``steps_per_trial`` a number of times for each batch size until one of the batch sizes generates and OOM
```

File: src/pytorch_lightning/callbacks/batch_size_finder.py [hunk 2]
```diff
@@ -26,9 +26,41 @@
...
+    error.
+
+    Args:
+        mode: search strategy to update the batch size:
+
+            - ``'power'``: Keep multiplying the batch size by 2, until we get an OOM error.
+            - ``'binsearch'``: Initially keep multiplying by 2 and after encountering an OOM error
+                do a binary search between the last successful batch size and the batch size that failed.
+
+        steps_per_trial: number of steps to run with a given batch size.
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:46]:
```suggestion
        steps_per_trial: Number of steps to run with a given batch size.
```

File: src/pytorch_lightning/callbacks/batch_size_finder.py [hunk 3]
```diff
@@ -26,9 +26,41 @@
...
+            Ideally 1 should be enough to test if a OOM error occurs,
+            however in practice a few are needed.
+
+        init_val: initial batch size to start the search with.
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:50]:
```suggestion
        init_val: Initial batch size to start the search with.
```

File: src/pytorch_lightning/callbacks/batch_size_finder.py [hunk 4]
```diff
@@ -26,9 +26,41 @@
...
+
+        max_trials: max number of increase in batch size done before
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:52]:
```suggestion
        max_trials: Maximum number of increases in batch size done before
```

File: src/pytorch_lightning/callbacks/batch_size_finder.py [hunk 5]
```diff
@@ -26,9 +26,41 @@
...
+            algorithm is terminated
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:53]:
```suggestion
            the algorithm is terminated
```

File: src/pytorch_lightning/callbacks/batch_size_finder.py [hunk 6]
```diff
@@ -26,9 +26,41 @@
...
+
+        batch_arg_name: name of the attribute that stores the batch size.
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:55]:
```suggestion
        batch_arg_name: Name of the attribute that stores the batch size.
```

File: src/pytorch_lightning/callbacks/lr_finder.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+from pytorch_lightning.utilities.exceptions import _TunerExitException, MisconfigurationException
+from pytorch_lightning.utilities.seed import isolate_rng
+
+
+class LearningRateFinder(Callback):
+    """The ``LearningRateFinder`` callback enables the user to do a range test of good initial learning rates, to
+    reduce the amount of guesswork in picking a good starting learning rate.
+
+    Args:
+        min_lr: minimum learning rate to investigate
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/lr_finder.py:34]:
```suggestion
        min_lr: Minimum learning rate to investigate
```

File: src/pytorch_lightning/callbacks/lr_finder.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+        max_lr: maximum learning rate to investigate
+
+        num_training: number of learning rates to test
+
+        mode: Search strategy to update learning rate after each batch:
+
+            - ``'exponential'`` (default): Will increase the learning rate exponentially.
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/lr_finder.py:42]:
```suggestion
            - ``'exponential'`` (default): Increases the learning rate exponentially.
```

File: src/pytorch_lightning/callbacks/lr_finder.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            - ``'linear'``: Will increase the learning rate linearly.
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/lr_finder.py:43]:
```suggestion
            - ``'linear'``: Increases the learning rate linearly.
```

File: src/pytorch_lightning/callbacks/lr_finder.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+        early_stop_threshold: threshold for stopping the search. If the
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/callbacks/lr_finder.py:45]:
```suggestion
        early_stop_threshold: Threshold for stopping the search. If the
```

File: src/pytorch_lightning/tuner/tuning.py [hunk 11]
```diff
@@ -231,7 +215,7 @@ def lr_find(
 
             mode: Search strategy to update learning rate after each batch:
 
-                - ``'exponential'`` (default): Will increase the learning rate exponentially.
+                - ``'exponential'``: Will increase the learning rate exponentially.
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/tuner/tuning.py:218]:
```suggestion
                - ``'exponential'``: Increases the learning rate exponentially.
```

File: src/pytorch_lightning/tuner/tuning.py [hunk 12]
```diff
@@ -231,7 +215,7 @@ def lr_find(
...
                 - ``'linear'``: Will increase the learning rate linearly.
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/tuner/tuning.py:219]:
```suggestion
                - ``'linear'``: Increases the learning rate linearly.
```

File: src/pytorch_lightning/tuner/tuning.py [hunk 13]
```diff
@@ -231,7 +215,7 @@ def lr_find(
...
 
             early_stop_threshold: threshold for stopping the search. If the
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/tuner/tuning.py:221]:
```suggestion
            early_stop_threshold: Threshold for stopping the search. If the
```

File: src/pytorch_lightning/callbacks/lr_finder.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+            or if you are using more than one optimizer.
+    """
+
+    SUPPORTED_MODES = ("linear", "exponential")
+
+    def __init__(
+        self,
+        min_lr: float = 1e-8,
+        max_lr: float = 1,
+        num_training: int = 100,
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/lr_finder.py:63]:
```suggestion
        num_training_batches: int = 100,
```

```suggestion
        num_training_steps: int = 100,
```

File: src/pytorch_lightning/callbacks/lr_finder.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+        self._min_lr = min_lr
+        self._max_lr = max_lr
+        self._num_training = num_training
+        self._mode = mode
+        self._early_stop_threshold = early_stop_threshold
+        self._update_attr = update_attr
+
+        self._early_exit = False
+        self.optimal_lr: Optional[_LRFinder] = None
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/lr_finder.py:80]:
```suggestion
        self.lr_finder: Optional[_LRFinder] = None
```

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 16]
```diff
@@ -274,29 +269,48 @@ def lr_find(
...
+
+    trainer.strategy.lr_scheduler_configs = []
+    trainer.strategy.optimizer_frequencies = []
     # avoid lr find being called multiple times
     trainer.auto_lr_find = False
     # Use special lr logger callback
     trainer.callbacks = [_LRCallback(num_training, early_stop_threshold, progress_bar_refresh_rate=1)]
     # No logging
-    trainer.loggers = [DummyLogger()] if trainer.loggers else []
+    trainer.logger = DummyLogger() if trainer.logger is not None else None
```
Comment (awaelchli) [inline src/pytorch_lightning/tuner/lr_finder.py:295]: Why wasn't the previous not working? The new implementation of trainer.logger should return the first logger in the list, which would be dummy logger here.

File: src/pytorch_lightning/tuner/tuning.py [hunk 17]
```diff
@@ -196,12 +177,15 @@ def scale_batch_size(
         self.trainer.auto_scale_batch_size = False
         return batch_size_finder.optimal_batch_size
 
+    # TODO: update docs
```
Comment (awaelchli) [inline src/pytorch_lightning/tuner/tuning.py:180]: Will you resolve this TODO in this PR? If not, the todo note should contain more information of what should be updated.

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Great work @rohitgr7
Team (@Borda @carmocca @tchaton @kaushikb11 etc. can we bring these two PRs in? They have been waiting for a long time. I know we are all not well familiar with the tuner, so let's just review as best as we can and let @rohitgr7 finalize the work?

**Comment by Borda** [comment]
> Great work @rohitgr7 Team (@Borda @carmocca @tchaton @kaushikb11 etc. can we bring these two PRs in? They have been waiting for a long time. I know we are all not well familiar with the tuner, so let's just review as best as we can and let @rohitgr7 finalize the work?

That would be GREAT!

**Review by Borda — APPROVED** [review]
Can we add docs on how to use this cb? :otter:

File: src/pytorch_lightning/callbacks/batch_size_finder.py (hunk 1 above)
Comment (otaj) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:36]:
```suggestion
    step function ``steps_per_trial`` times for each batch size until one of the batch size generates an OOM
```

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/callbacks/lr_finder.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+                min_lr=self._min_lr,
+                max_lr=self._max_lr,
+                num_training=self._num_training,
+                mode=self._mode,
+                early_stop_threshold=self._early_stop_threshold,
+                update_attr=self._update_attr,
+            )
+
+        if self._early_exit:
+            raise _TunerExitException()
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/lr_finder.py:96]: Looks to me like this PR is waiting on #11089 to be merged. As you didn't add the logic to catch this here. Are they meant to be sequenced?

Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/lr_finder.py:96]: yes. check the base :)

**Comment by Borda** [comment]
@rohitgr7, mind checking the collisions? :otter:

**Pull request merged**